paprika/build.py

```diff
@@ -1,185 +1,141 @@
-                            ". Only 0, 1, 2, and None are valid")
-        if neutralize:
-            f.write("addionsrand {} {} 0\n".format(unit, counter_cation))
-            f.write("addionsrand {} {} 0\n".format(unit, counter_anion))
-        if addion_residues:
-            for i, res in enumerate(addion_residues):
-                f.write("addionsrand {} {} {}\n".format(
-                    unit, res, addion_values[i]))
-        if removewat:
-            for watnum in removewat:
-                f.write("remove {} {}.{}\n".format(unit, unit, watnum))
-        if saveprefix:
-            f.write("savepdb {} {}.pdb\n".format(unit, saveprefix))
-            f.write("saveamberparm {} {}.prmtop {}.rst7\n".format(
-                unit, saveprefix, saveprefix))
-        f.write("desc {}\n".format(unit))
+            raise Exception(
+                "Incorrect pbctype value provided: " + str(options['pbc_type']) + ". Only 0, 1, 2, and None are valid")
+        if options['neutralize']:
+            f.write("addionsrand {} {} 0\n".format(options['unit'], options['counter_cation']))
+            f.write("addionsrand {} {} 0\n".format(options['unit'], options['counter_anion']))
+        # Additional ions should be specified as a list, with residue name and number of ions in pairs, like ['Na+',
+        # 5] for five additional sodium ions. By this point, if the user specified a molality or molarity,
+        # it should already have been converted into a number.
+        if options['add_ion_residues']:
+            for residue, amount in zip(options['add_ion_residues'][0::2],
+                                       options['add_ion_residues'][1::2]):
+                f.write("addionsrand {} {} {}\n".format(options['unit'], residue, amount))
+        if options['remove_water']:
+            for water_number in options['remove_water']:
+                f.write("remove {} {}.{}\n".format(options['unit'], options['unit'], water_number))
+
+        if options['output_prefix']:
+            f.write("savepdb {} {}.pdb\n".format(options['unit'], options['output_prefix']))
+            f.write("saveamberparm {} {}.prmtop {}.rst7\n".format(options['unit'], options['output_prefix'],
+                                                                  options['output_prefix']))
+        else:
+            pass
+        f.write("desc {}\n".format(options['unit']))
         f.write("quit\n")
 
 
-def run_tleap(filename, directory):
+def run_tleap(path='./', file_name='tleap.in'):
     """
     Execute tLEaP.
     """
 
-    check_for_leap_log(path=directory)
-    p = sp.Popen(
-        'tleap -s -f ' + filename,
-        stdout=sp.PIPE,
-        bufsize=1,
-        universal_newlines=True,
-        cwd=directory,
-        shell=True)
+    utils.check_for_leap_log(path=path)
+    p = sp.Popen('tleap -s -f ' + file_name, stdout=sp.PIPE, bufsize=1, universal_newlines=True, cwd=path, shell=True)
```

Comment:
Do we need to use shell=True?  I think we should avoid that ... just make an exec_string.

Reply:
Fixed https://github.com/slochower/pAPRika/pull/43/commits/924df83e7ad32848955bdf82ab98c74faadefe89